server/src/main/java/org/elasticsearch/search/aggregations/bucket/histogram/AutoDateHistogramAggregator.java

```diff
@@ -99,95 +206,219 @@ public DeferringBucketCollector getDeferringCollector() {
+                if (bucketOrd < 0) { // already seen
+                    bucketOrd = -1 - bucketOrd;
+                    collectExistingBucket(sub, doc, bucketOrd);
+                    return roundingIdx;
+                }
+                collectBucket(sub, doc, bucketOrd);
+                liveBucketCountUnderestimate = context.bigArrays().grow(liveBucketCountUnderestimate, owningBucketOrd + 1);
+                int estimatedBucketCount = liveBucketCountUnderestimate.increment(owningBucketOrd, 1);
+                return increaseRounding(owningBucketOrd, estimatedBucketCount, rounded, roundingIdx);
+            }
+
+            /**
+             * Increase the rounding of {@code owningBucketOrd} using
+             * estimated, bucket counts, {@link #rebucket() rebucketing} the all
+             * buckets if the estimated number of wasted buckets is too high.
+             */
+            private byte increaseRounding(long owningBucketOrd, int oldEstimatedBucketCount, long newKey, byte oldRounding) {
+                if (oldRounding >= roundingInfos.length - 1) {
+                    return oldRounding;
+                }
+                if (bounds.size() < owningBucketOrd * 2 + 2) {
+                    long oldSize = bounds.size();
+                    bounds = context.bigArrays().grow(bounds, owningBucketOrd * 2 + 2);
+                    for (long b = oldSize; b < bounds.size(); b++) {
+                        bounds.set(b, (b & 1L) == 0L ? Long.MAX_VALUE : Long.MIN_VALUE);
                     }
-                    bucketOrds = newBucketOrds;
                 }
+                long min = Math.min(bounds.get(owningBucketOrd * 2), newKey);
+                bounds.set(owningBucketOrd * 2, min);
+                long max = Math.max(bounds.get(owningBucketOrd * 2 + 1), newKey);
+                bounds.set(owningBucketOrd * 2 + 1, max);
+                if (oldEstimatedBucketCount <= targetBuckets * roundingInfos[oldRounding].getMaximumInnerInterval()
+                        && max - min <= targetBuckets * roundingInfos[oldRounding].getMaximumRoughEstimateDurationMillis()) {
+                    return oldRounding;
+                }
+                long oldRoughDuration = roundingInfos[oldRounding].roughEstimateDurationMillis;
+                byte newRounding = oldRounding;
+                int newEstimatedBucketCount;
+                do {
+                    newRounding++;
+                    double ratio = (double) oldRoughDuration / (double) roundingInfos[newRounding].getRoughEstimateDurationMillis();
+                    newEstimatedBucketCount = (int) Math.ceil(oldEstimatedBucketCount * ratio);
+                } while (newRounding < roundingInfos.length - 1 && (
+                    newEstimatedBucketCount > targetBuckets * roundingInfos[newRounding].getMaximumInnerInterval()
+                        || max - min > targetBuckets * roundingInfos[newRounding].getMaximumRoughEstimateDurationMillis()));
```

Comment:
should there be unit tests for when:

```
(max - min > targetBuckets * roundingInfos[newRounding].getMaximumRoughEstimateDurationMillis()) == false
&& (newEstimatedBucketCount > targetBuckets * roundingInfos[newRounding].getMaximumInnerInterval()))
```
?

Reply:
We do cover this case in the test case, but there isn't a unit test for this exactly. do you think it is worth pulling this out somehow?